src/vs/workbench/parts/terminal/browser/terminalQuickOpen.ts

```diff
@@ -0,0 +1,137 @@
+		private open: () => void
+	) {
+		super();
+	}
+
+	public getLabel(): string {
+		return this.label;
+	}
+
+	public getCategory(): string {
+		return this.category;
+	}
+
+	public getAriaLabel(): string {
+		return nls.localize('entryAriaLabel', "{0}, terminal picker", this.getLabel());
+	}
+
+	public run(mode: Mode, context: IEntryRunContext): boolean {
+		if (mode === Mode.OPEN) {
+			return this.runOpen(context);
+		}
+
+		return super.run(mode, context);
+	}
+
+	private runOpen(context: IEntryRunContext): boolean {
+		setTimeout(() => {
+			this.open();
+		}, 0);
+
+		return true;
+	}
+}
+
+export class TerminalPickerHandler extends QuickOpenHandler {
+
+	constructor(
+		@ITerminalService private terminalService: ITerminalService,
+		@IPanelService private panelService: IPanelService
+	) {
+		super();
+	}
+
+	public getResults(searchValue: string): TPromise<QuickOpenModel> {
+		searchValue = searchValue.trim();
+		const normalizedSearchValueLowercase = strings.stripWildcards(searchValue).toLowerCase();
+
+		const terminalEntries = this.getTerminals();
+
+		const entries = terminalEntries.filter(e => {
+			if (!searchValue) {
+				return true;
+			}
+
+			if (!scorer.matches(e.getLabel(), normalizedSearchValueLowercase) && !scorer.matches(e.getCategory(), normalizedSearchValueLowercase)) {
+				return false;
+			}
+
+			const { labelHighlights, descriptionHighlights } = QuickOpenEntry.highlight(e, searchValue);
+			e.setHighlights(labelHighlights, descriptionHighlights);
+
+			return true;
+		});
+
+		let lastCategory: string;
+		entries.forEach((e, index) => {
+			if (lastCategory !== e.getCategory()) {
+				lastCategory = e.getCategory();
+
+				e.setShowBorder(index > 0);
+				e.setGroupLabel(lastCategory);
+			} else {
+				e.setShowBorder(false);
+				e.setGroupLabel(void 0);
+			}
+		});
+
+		return TPromise.as(new QuickOpenModel(entries));
+	}
+
+	private getTerminals(): TerminalEntry[] {
+		const termninalEntries: TerminalEntry[] = [];
+		const taskEntries: TerminalEntry[] = [];
+		const terminals = this.terminalService.terminalInstances;
+
+		terminals.forEach((terminal, index) => {
+			const terminalsCategory = nls.localize('terminals', "Terminal");
+			const taskCategory = nls.localize('task', "Task");
+			if (terminal.title.indexOf('Task') >= 0) {
+				taskEntries.push(new TerminalEntry(nls.localize('terminalTitle', "{0}: {1}", index + 1, terminal.title), taskCategory, () => {
+					this.terminalService.showPanel(true).done(() => {
+						this.terminalService.setActiveInstance(terminal);
+					}, errors.onUnexpectedError);
+				}));
+			} else {
+				termninalEntries.push(new TerminalEntry(nls.localize('terminalTitle', "{0}: {1}", index + 1, terminal.title), terminalsCategory, () => {
+					this.terminalService.showPanel(true).done(() => {
+						this.terminalService.setActiveInstance(terminal);
+					}, errors.onUnexpectedError);
+				}));
+			}
+
+
+		});
+		return termninalEntries.concat(taskEntries);
```

Comment:
Part of me likes this since the original idea was that mixing user terminals and task terminals together is confusing. What if they were separated in the dropdown?